lib/types.nix

```diff
@@ -467,15 +467,10 @@ rec {
       let
         inherit (lib.modules) evalModules;
 
-        coerce = unify: value: if isFunction value
-          then setFunctionArgs (args: unify (value args)) (functionArgs value)
-          else unify (if shorthandOnlyDefinesConfig then { config = value; } else value);
-
         allModules = defs: modules ++ imap1 (n: { value, file }:
-          if isAttrs value || isFunction value then
-            # Annotate the value with the location of its definition for better error messages
-            coerce (lib.modules.unifyModuleSyntax file "${toString file}-${toString n}") value
-          else value
+          if isAttrs value && shorthandOnlyDefinesConfig
+          then { _file = file; config = value; }
+          else { _file = file; imports = [ value ]; }
```

Comment:
- `n` is now unused.
 - Usually an extra `imports` indirection doesn't matter, but some error messages print a path through the import graph. Those will be a bit harder to understand after this change

So it doesn't seem to be purely refactoring.

In all fairness those messages aren't great, but this change is a small step back.

Perhaps we could revisit those error messages and make them report an imports "trace" with actual file/line/column positions.

Reply:
Boy was I wrong.

This seems to be a solution to the same problem that https://github.com/NixOS/nixpkgs/pull/177157 solves.

> * `n` is now unused.

`n` was a technicality of a previous solution for adding `_file`. Nothing more, nothing less.

> Usually an extra imports indirection doesn't matter, but some error messages print a path through the import graph.

I don't think these occur frequently, and if they do they're a problem that can be solved separately. Adding more `_file`, like this code does, may prevent such messages in the first place.

------

off-topic section :turtle: 

> So it doesn't seem to be purely refactoring.

Picking nits...

> Perhaps we could revisit those error messages and make them report an imports "trace" with actual file/line/column positions.

Something with `unsafeGetAttrPos` seems like a better path for improving location reporting.